tests/beta/tapes/test_tape.py

```diff
@@ -0,0 +1,667 @@
+
+        for pinfo, pval in zip(tape._par_info.values(), new_params):
+            assert pinfo["op"].data[pinfo["p_idx"]] == pval
+
+        assert tape.get_parameters(free_only=False) == new_params
+
+    def test_setting_parameters_error(self, make_tape):
+        """Test that exceptions are raised if incorrect parameters
+        are attempted to be set"""
+        tape, _ = make_tape
+
+        with pytest.raises(ValueError, match="Number of provided parameters invalid"):
+            tape.set_parameters([0.54])
+
+        with pytest.raises(ValueError, match="Number of provided parameters invalid"):
+            tape.trainable_params = {2, 3}
+            tape.set_parameters([0.54, 0.54, 0.123])
+
+    def test_array_parameter(self):
+        """Test that array parameters integrate properly"""
+        a = np.array([1, 1, 0, 0]) / np.sqrt(2)
+        params = [a, 0.32, 0.76, 1.0]
+
+        with QuantumTape() as tape:
+            op = qml.QubitStateVector(params[0], wires=0)
+            qml.Rot(params[1], params[2], params[3], wires=0)
+
+        assert tape.num_params == len(params)
+        assert tape.get_parameters() == params
+
+        b = np.array([0, 1, 0, 0])
+        new_params = [b, 0.543, 0.654, 0.123]
+        tape.set_parameters(new_params)
+        assert tape.get_parameters() == new_params
+
+        assert np.all(op.data[0] == b)
+
+    def test_measurement_parameter(self):
+        """Test that measurement parameters integrate properly"""
+        H = np.array([[1, 0], [0, -1]])
+        params = [0.32, 0.76, 1.0, H]
+
+        with QuantumTape() as tape:
+            qml.Rot(params[0], params[1], params[2], wires=0)
+            obs = qml.Hermitian(params[3], wires=0)
+            expval(obs)
+
+        assert tape.num_params == len(params)
+        assert tape.get_parameters() == params
+
+        H2 = np.array([[0, 1], [1, 1]])
+        new_params = [0.543, 0.654, 0.123, H2]
+        tape.set_parameters(new_params)
+        assert tape.get_parameters() == new_params
+
+        assert np.all(obs.data[0] == H2)
+
+
+class TestInverse:
+    """Tests for tape inversion"""
+
+    def test_inverse(self):
+        """Test that inversion works as expected"""
+        init_state = np.array([1, 1])
+        p = [0.1, 0.2, 0.3, 0.4]
+
+        with QuantumTape() as tape:
+            prep = qml.BasisState(init_state, wires=[0, "a"])
+            ops = [qml.RX(p[0], wires=0), qml.Rot(*p[1:], wires=0).inv(), qml.CNOT(wires=[0, "a"])]
+            m1 = probs(wires=0)
+            m2 = probs(wires="a")
+
+        tape.inv()
+
+        # check that operation order is reversed
+        assert tape.operations == [prep] + ops[::-1]
+
+        # check that operations are inverted
+        assert ops[0].inverse
+        assert not ops[1].inverse
+        assert ops[2].inverse
+
+        # check that parameter order has reversed
+        assert tape.get_parameters() == [init_state, p[1], p[2], p[3], p[0]]
```

Comment:
Nice edge case you're checking here :+1:

Reply:
this was a tough one to get working 😬